src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/LicenceService.kt

```diff
@@ -0,0 +1,69 @@
+package uk.gov.justice.digital.hmpps.hmppshdcapi.licences
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppshdcapi.model.HdcLicence
+
+@Service
+class LicenceService(
+  private val licenceRepository: LicenceRepository,
+  private val objectMapper: ObjectMapper,
+) {
+  fun getByBookingId(bookingId: Long): HdcLicence? {
+    val licenceObject = licenceRepository.findLicenceByBookingId(bookingId) ?: return null
+
+    val licence = licenceObject.licence
+
+    if (licence.isNullOrEmpty()) {
+      return null
+    }
+
+    val cas2ReferralObject = licence["bassReferral"]
+    val cas2Referral = objectMapper.convertValue(cas2ReferralObject, Cas2Referral::class.java)
+
+    val curfewObject = licence["curfew"]
+    val curfew = objectMapper.convertValue(curfewObject, Curfew::class.java)
+
+    val proposedAddressObject = licence["proposedAddress"]
+    val proposedAddress = objectMapper.convertValue(proposedAddressObject, ProposedAddress::class.java)
+
+    return transformToHdcLicence(
+      curfewAddress = getAddress(curfew, cas2Referral, proposedAddress),
+      firstNightCurfewHours = curfew.firstNight,
+      curfewHours = curfew.curfewHours,
+    )
+  }
+
+  fun getAddress(curfew: Curfew, cas2Referral: Cas2Referral, proposedAddress: ProposedAddress): String {
+    val isCurfewApprovedPremisesRequired = curfew.approvedPremises?.required == Decision.Yes
+    val isCas2ApprovedPremisesRequired = cas2Referral.bassAreaCheck?.approvedPremisesRequiredYesNo == Decision.Yes
+    val isCas2Requested = cas2Referral.bassRequest.bassRequested == Decision.Yes
+    val isCas2Accepted = cas2Referral.bassOffer?.bassAccepted == OfferAccepted.Yes
+
+    if (isCurfewApprovedPremisesRequired && !isCas2ApprovedPremisesRequired) {
+      return formatAddress(curfew.approvedPremisesAddress!!)
+    }
+
+    if (isCas2ApprovedPremisesRequired) {
+      return formatAddress(cas2Referral.approvedPremisesAddress!!)
+    }
+
+    if (isCas2Requested && isCas2Accepted) {
+      val cas2Address = cas2Referral.bassOffer as Cas2Offer
+      val address = Address(
+        addressLine1 = cas2Address.addressLine1,
+        addressLine2 = cas2Address.addressLine2,
+        addressTown = cas2Address.addressTown,
+        postCode = cas2Address.postCode,
+      )
+      return formatAddress(address)
+    }
+    return formatAddress(proposedAddress.curfewAddress!!)
+  }
+
+  private fun formatAddress(addressObject: Address): String = if (addressObject.addressLine2 != null) {
+    "${addressObject.addressLine1}, ${addressObject.addressLine2}, ${addressObject.addressTown}, ${addressObject.postCode}"
+  } else {
+    "${addressObject.addressLine1}, ${addressObject.addressTown}, ${addressObject.postCode}"
+  }
```

Comment:
**suggestion:** it might be more flexible to expose a rich data object rather than one string via the API.

Reply:
We had a chat about this and thought that Craig mentioned having the address as a single string as opposed to an object following an initial design conversation around AfER. Happy to change it to an object as can't really remember the conversation so can easily change at a later stage.